{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import os\n",
    "import pickle\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "input_paths = [\n",
    "    '/MMCI/TM/scratch/pebert/diploid_assembly/output/fastq_validation',\n",
    "    '/MMCI/TM/scratch/pebert/diploid_assembly/side_tracks/read_stats'\n",
    "]\n",
    "input_files = []\n",
    "for p in input_paths:\n",
    "    files_under_path = [os.path.join(p, f) for f in os.listdir(p) if f.endswith('.pck')]\n",
    "    input_files.extend(files_under_path)\n",
    "\n",
    "output_root = '/MMCI/TM/scratch/pebert/diploid_assembly/output/plots/fastq_stats'\n",
    "\n",
    "def plot_gc_content(title, binned_values):\n",
    "    \n",
    "    fig, ax = plt.subplots(figsize=(6,4))\n",
    "    bins = np.array(list(binned_values.keys()), dtype=np.int8)\n",
    "    bins.sort()\n",
    "    counts = np.array([binned_values[b] for b in bins], dtype=np.int32)\n",
    "    total_read_count = counts.sum()\n",
    "    \n",
    "    ax.bar(bins, counts, width=1)\n",
    "    ax.set_xlabel('G+C content (% ; binned)', fontsize=12)\n",
    "    ax.set_ylabel('Read count', fontsize=12)\n",
    "    ax.text(0.7, 0.8, s='Total reads:\\n{}'.format(total_read_count),\n",
    "            transform=ax.transAxes, fontdict={'fontsize': 12})\n",
    "    tt = ax.set_title(title, fontsize=12)\n",
    "    tt.set_position([0.5, 1.2])\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    return fig, [tt]\n",
    "    \n",
    "\n",
    "def plot_read_length_distribution(title, read_lengths):\n",
    "    gsize = 3 * 1e9\n",
    "    length_values = np.array(list(read_lengths.keys()), dtype=np.int32)\n",
    "    length_values.sort()\n",
    "    read_counts = np.array([read_lengths[l] for l in length_values])\n",
    "    base_counts = np.array([read_lengths[l] * l for l in length_values])\n",
    "    \n",
    "#     total_read_length = 0\n",
    "#     ul_read_length = 0\n",
    "#     for length, count in read_lengths.items():\n",
    "#         total_read_length += length * count\n",
    "#         if length >= 100000:\n",
    "#             ul_read_length += length * count\n",
    "    \n",
    "    base_count_bins = []\n",
    "    read_count_bins = []\n",
    "    lower = [0, 1000] + list(range(5000, 105000, 5000))\n",
    "    higher = lower[1:] + [10000000]\n",
    "    \n",
    "    for l, h in zip(lower, higher):\n",
    "        select_low = np.array(length_values >= l, dtype=np.bool)\n",
    "        select_high = np.array(length_values < h, dtype=np.bool)\n",
    "        selector = np.logical_and(select_low, select_high)\n",
    "        base_count_bins.append(base_counts[selector].sum())\n",
    "        read_count_bins.append(read_counts[selector].sum())\n",
    "    \n",
    "    fig, ax = plt.subplots(figsize=(4,3))\n",
    "    x_pos = list(range(len(lower)))\n",
    "    x_labels = list(map(str, [x//1000 for x in lower[1:]]))\n",
    "    x_labels = [x if i % 3 == 0 else ' ' for i, x in enumerate(x_labels)]\n",
    "    x_labels.append('UL')\n",
    "    \n",
    "    ax.bar(x_pos, base_count_bins, width=0.8, align='center')\n",
    "    ax.set_xlabel('Read length < X kbp', fontsize=12)\n",
    "    ax.set_xticks(x_pos)\n",
    "    ax.set_xticklabels(x_labels, fontsize=10)\n",
    "    ax.set_ylabel('Base count', fontsize=12)\n",
    "    \n",
    "    total_read_count = sum(read_count_bins)\n",
    "    total_base_pair = sum(base_count_bins)\n",
    "    \n",
    "    ultralong_read_count = read_count_bins[-1]\n",
    "    ultralong_base_pair = base_count_bins[-1]\n",
    "    \n",
    "    #assert ultralong_base_pair == ul_read_length\n",
    "    #assert total_base_pair == total_read_length\n",
    "    \n",
    "    text = 'Reads: {}\\n'.format(total_read_count)\n",
    "    text += ' ~ {} Gbp\\n'.format(round(total_base_pair / 1e9, 1))\n",
    "    text += ' ~ {}x\\n'.format(round(total_base_pair / gsize, 2))\n",
    "    text += 'UL: {}\\n'.format(ultralong_read_count)\n",
    "    text += ' ~ {} Gbp\\n'.format(round(ultralong_base_pair / 1e9, 1))\n",
    "    text += ' ~ {}x'.format(round(ultralong_base_pair / gsize, 2))\n",
    "    \n",
    "\n",
    "    txt = ax.text(1, 0.4, s=text, transform=ax.transAxes, fontdict={'fontsize': 12})\n",
    "    tt = ax.set_title(title, fontsize=12)\n",
    "    tt.set_position([0.5, 1.2])\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    return fig, [txt, tt]\n",
    "        \n",
    "\n",
    "def plot_sequenced_bases(title, base_count):\n",
    "    \n",
    "    fig, ax = plt.subplots(figsize=(4,3))\n",
    "    \n",
    "    count = [base_count[b] for b in 'ACGT']\n",
    "    xpos = list(range(4))\n",
    "    ax.bar(xpos, count, width=0.8, align='center')\n",
    "    ax.set_xticks(xpos)\n",
    "    ax.set_xticklabels(['A', 'C', 'G', 'T'], fontsize=10)\n",
    "    ax.set_xlabel('Nucleotide', fontsize=12)\n",
    "    ax.set_ylabel('# sequenced', fontsize=12)\n",
    "    tt = ax.set_title(title, fontsize=12)\n",
    "    tt.set_position([0.5, 1.2])\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    return fig, [tt]\n",
    "\n",
    "\n",
    "title_filetype = {'cmp': 'Raw',\n",
    "                 'tag-h1': 'Hap1',\n",
    "                 'tag-h2': 'Hap2',\n",
    "                 'tag-un': 'Untag',\n",
    "                 'tag-h1-un': 'Hap1-Un',\n",
    "                 'tag-h2-un': 'Hap2-Un'}\n",
    "\n",
    "prettify_sample = {\n",
    "    'HG00514.sequel2.pb-ccs': 'HG.514 PB-Sq2 CCS',\n",
    "    'HG00514.sequel2.pb-clr': 'HG.514 PB-Sq2 CLR',\n",
    "    'HG00514.sequel2.pb-clr-25': 'HG.514 PB-Sq2 CLR (0.25)'\n",
    "}\n",
    "\n",
    "plot_gc_cont = True\n",
    "plot_read_length = True\n",
    "plot_seq_bases = True\n",
    "\n",
    "save_figures = True\n",
    "\n",
    "for infile in input_files:\n",
    "    filename = os.path.basename(infile)\n",
    "    if 'HG00514' not in filename:\n",
    "        continue\n",
    "    if 'EDEVI' in infile:\n",
    "        filetype = filename.rsplit('.', 1)[0].rsplit('_', 2)[-3:]\n",
    "        filetype = '_'.join(filetype)\n",
    "        filetype = filetype.replace('.', '_')\n",
    "        if 'subreads' in filename or filename.endswith('.ccs.pck'):\n",
    "            plot_filetype = 'HG.514 PB-Sq2 subreads'\n",
    "            sample = filename.split('.')[2]\n",
    "            plot_sample = filename.split('.')[2]\n",
    "    else:\n",
    "        filetype = infile.rsplit('.', 3)[-3]\n",
    "        plot_filetype = title_filetype[filetype]\n",
    "        sample = os.path.basename(infile).split('.')[:3]\n",
    "        sample = '.'.join(sample)\n",
    "        \n",
    "    plot_sample = prettify_sample.get(sample, sample)\n",
    "        \n",
    "    dump = pickle.load(open(infile, 'rb'))\n",
    "    \n",
    "    outpath = os.path.join(output_root, sample, '{}_{}_'.format(sample, filetype.lower()))\n",
    "    os.makedirs(os.path.dirname(outpath), exist_ok=True)\n",
    "    if plot_gc_cont:\n",
    "        fig, exart = plot_gc_content('{}\\n{}'.format(plot_filetype, plot_sample), dump['gc_bins'])\n",
    "        if save_figures:\n",
    "            fig.savefig(outpath + 'gc-cont.png', dpi=300, bbox_inches='tight', extra_artists=exart)\n",
    "        plt.close(fig)\n",
    "    \n",
    "    if plot_read_length:\n",
    "        fig, exart = plot_read_length_distribution('{}\\n{}'.format(plot_filetype, plot_sample), dump['len_stats'])\n",
    "        if save_figures:\n",
    "            fig.savefig(outpath + 'read-len.png', dpi=300, bbox_inches='tight', extra_artists=exart)\n",
    "        plt.close(fig)\n",
    "    \n",
    "    if plot_seq_bases:\n",
    "        fig, exart = plot_sequenced_bases('{}\\n{}'.format(plot_filetype, plot_sample), dump['nuc_stats'])\n",
    "        if save_figures:\n",
    "            fig.savefig(outpath + 'seq-bases.png', dpi=300, bbox_inches='tight', extra_artists=exart)\n",
    "        plt.close(fig)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
